### PR TITLE
Feature: Update view model to support sections

### PR DESCRIPTION
**Pull request by billypchan**
## What's new in this PR?

Rewrite the conversation list view model for dynamic sections.

### TODO
- [x] remove dummy toggle folder button
- [x] test for ConversationListViewModel exposed methods
- [ ] use diff kit to compare `[ZMConversation]`
- [x] move observer of `ZMConversation` to Conversation cell

### dependency 
- [x] https://github.com/wireapp/wire-ios-data-model/pull/782

### CODE REVIEW

File: Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/Model/ConversationListViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,463 @@
...
+            }
+        }
+
+        return nil
+    }
+
+    private static func newList(for section: Section) -> [AnyHashable] {
+        guard let userSession = ZMUserSession.shared() else { return [] }
+
+        switch section {
```
Comment (typfel) [inline Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/Model/ConversationListViewModel.swift]: Why not use `userSession.conversationDirectory.conversation(by: .pending)` in this switch

File: Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/Model/ConversationListViewModel.swift [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+        case contacts
+
+        /// gorup conversations
+        case group
+
+        ///TODO:
+        //    case customFolder(folder: FolderType)
+    }
+
+    private struct Folder {
```
Comment (typfel) [inline Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/Model/ConversationListViewModel.swift]: Why don't we name this `Section` since that's more closely matches what's this struct is used for.

File: Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/Model/ConversationListViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+import Foundation
+
+// Placeholder for conversation requests item
+///TODO: create a protocol, shared with ZMConversation
+@objc
+final class ConversationListConnectRequestsItem : NSObject {}
+
+
+final class ConversationListViewModel: NSObject {
+    enum Section: CaseIterable {
```
Comment (typfel) [inline Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/Model/ConversationListViewModel.swift]: This could move inside the `struct Folder` and maybe just be called `Type` unless that's a reserved word.

File: Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/Model/ConversationListViewModel.swift [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+        /// gorup conversations
+        case group
+
+        ///TODO:
+        //    case customFolder(folder: FolderType)
+    }
+
+    private struct Folder {
+        var section: Section
+        var items: [AnyHashable]
```
Comment (typfel) [inline Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/Model/ConversationListViewModel.swift]: Would be cool to have the items strongly typed but maybe let's do in separate PR if this is too big of change.

File: Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/Model/ConversationListViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+    @objc
+    func numberOfItems(inSection sectionIndex: UInt) -> UInt {
+        guard sectionIndex < sectionCount else { return 0 }
+
+        return UInt(folders[Int(sectionIndex)].items.count)
+    }
+
+    private
+    func numberOfItems(in section: Section) -> Int? {
+        for (folder) in folders {
```
Comment (typfel) [inline Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/Model/ConversationListViewModel.swift]: This could be shortened with a `folders.first(where: { $0.section == section }).items.count ?? nil`

File: Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/Model/ConversationListViewModel.swift (hunk 4 above)
Comment (billypchan) [reply Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/Model/ConversationListViewModel.swift]: The member of `items` would be either `ConversationListConnectRequestsItem` or `ZMConversation`. I would change it to a common protocol type after refactored the view controller.

File: Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/Model/ConversationListViewModel.swift (hunk 1 above)
Comment (billypchan) [reply Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/Model/ConversationListViewModel.swift]: Will change other cases also.

File: Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/Model/ConversationListViewModel.swift [hunk 6]
```diff
@@ -162,7 +155,7 @@ final class ConversationListViewModel: NSObject {
 
         switch section {
         case .contactRequests:
-            return  ZMConversationList.pendingConnectionConversations(inUserSession: userSession).count > 0 ? [contactRequestsItem] : []
+            return userSession.managedObjectContext.conversationListDirectory().conversations(by: .pending).count > 0 ? [contactRequestsItem] : []
```
Comment (typfel) [inline Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/Model/ConversationListViewModel.swift]: We should not access `ZMConversationListDirectory` directly through `userSession.managedObjectContext.conversationListDirectory. It's exposed as protocol directly on the user session in this PR https://github.com/wireapp/wire-ios-sync-engine/pull/1087

Comment (billypchan) [reply Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/Model/ConversationListViewModel.swift]: done

File: Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/Model/ConversationListViewModel.swift (hunk 3 above)
Comment (billypchan) [reply Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/Model/ConversationListViewModel.swift]: rename to `Kind` to prevent confliction
